test/new-e2e/utils/e2e/e2e.go

```diff
@@ -3,8 +3,220 @@
+//
+// The third parameter defines the environment. See "Using existing stack definition" section for more information about a environment definition.
+//
+// 3. Write a test function
+//
+//	func (v *vmSuite) TestBasicVM() {
+//		v.Env().VM.Execute("ls")
+//	}
+//
+// [e2e.Suite.Env] gives access to the components in your environment.
+//
+// Depending on your stack definition, [e2e.Suite.Env] can provide the following objects:
+//   - [client.VM]: A virtual machine where you can execute commands.
+//   - [client.Agent]: A struct that provides methods to run datadog agent commands.
+//   - [client.Fakeintake]: A struct that provides methods to run queries to a fake instance of Datadog intake.
+//
+// # Using an existing stack definition
+//
+// The stack definition defines the components available in your environment.
+//
+//	type vmSuite struct {
+//		e2e.Suite[e2e.VMEnv]
+//	}
+//
+//	func TestVMSuite(t *testing.T) {
+//		e2e.Run(t, &vmSuite{}, e2e.EC2VMStackDef())
+//	}
+//
+// In this example, the components available are defined by the struct [e2e.VMEnv] which contains a virtual machine.
+// The generic type of [e2e.Suite] must match the type of the stack definition.
+// In our example, [e2e.EC2VMStackDef] returns an instance of [*e2e.StackDefinition][[e2e.VMEnv]].
+//
+// The following default stack definitions are provided:
+//   - [e2e.EC2VMStackDef] creates an environment with a virtual machine. See [e2e.EC2VMStackDef] for more information about the supported options.
+//   - [e2e.AgentStackDef] creates an environment with an Agent installed on a virtual machine. See [e2e.AgentStackDef] for more information about the supported options.
+//
+// # Defining your stack definition
+//
+// In some special cases, you have to define a custom environment.
+// Here is an example of an environment with Docker installed on a virtual machine.
+//
+//	type dockerSuite struct {
+//		e2e.Suite[e2e.VMEnv]
+//	}
+//
+//	func TestDockerSuite(t *testing.T) {
+//		e2e.Run(t, &dockerSuite{}, e2e.EnvFactoryStackDef(dockerEnvFactory))
+//	}
+//
+//	func dockerEnvFactory(ctx *pulumi.Context) (*e2e.VMEnv, error) {
+//		vm, err := ec2vm.NewUnixEc2VM(ctx)
+//		if err != nil {
+//			return nil, err
+//		}
+//
+//		_, err = docker.NewAgentDockerInstaller(vm.UnixVM, docker.WithAgent(docker.WithAgentImageTag("7.42.0")))
+//
+//		if err != nil {
+//			return nil, err
+//		}
+//
+//		return &e2e.VMEnv{
+//			VM: client.NewVM(vm),
+//		}, nil
+//	}
+//
+//	func (docker *dockerSuite) TestDocker() {
+//		docker.Env().VM.Execute("docker container ls")
+//	}
+//
+// [e2e.EnvFactoryStackDef] is used to define a custom environment.
+// Here is a non exhaustive list of components that can be used to create a custom environment:
+//   - [EC2 VM]: Provide methods to create a virtual machine on EC2.
+//   - [Agent]: Provide methods to install the Agent on a virtual machine
+//   - [File Manager]: Provide methods to manipulate files and folders
+//
+// # Organizing your tests
```

Comment:
I'd add a **NOTE** (in bold) to emphasize that `TestXX` on `Suite` are **not** executed in top->bottom order (unlike normal go tests) and that test ordering is random and not guaranteed.